# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: netwerk/base/BackgroundFileSaver.cpp
# Commit: af7b5fa1de26
# Full Hash: af7b5fa1de262d973d90b400bb568a47bceccad0
# Author: Chris Fronk <fronkc1@gmail.com>
# Date: 2020-06-10 14:02:18
# Description:
#   Bug 1644237 - Fix crash in BackgroundFileSaver r=valentin,necko-reviewers
#   
#   Previous implementation on Finish thread was shutdown but not set to
#   null, as background task queue cannot be shutdown need to add nullptr
#   check around access in notify worker.
# ==============================================================================

diff -r f399bb3b6b1b -r af7b5fa1de26 netwerk/base/BackgroundFileSaver.cpp
--- a/netwerk/base/BackgroundFileSaver.cpp	Wed Jun 10 05:52:53 2020 +0000
+++ b/netwerk/base/BackgroundFileSaver.cpp	Wed Jun 10 06:23:58 2020 +0000
@@ -286,12 +286,20 @@
   }
 
   if (!mAsyncCopyContext) {
+    // Background event queues are not shutdown and could be called after
+    // the queue is reset to null.  To match the behavior of nsIThread
+    // return NS_ERROR_UNEXPECTED
+    if (!mBackgroundET) {
+      return NS_ERROR_UNEXPECTED;
+    }
+
     // Copy is not in progress, post an event to handle the change manually.
     rv = mBackgroundET->Dispatch(
         NewRunnableMethod("net::BackgroundFileSaver::ProcessAttention", this,
                           &BackgroundFileSaver::ProcessAttention),
         NS_DISPATCH_EVENT_MAY_BLOCK);
     NS_ENSURE_SUCCESS(rv, rv);
+
   } else if (aShouldInterruptCopy) {
     // Interrupt the copy.  The copy will be resumed, if needed, by the
     // ProcessAttention function, invoked by the AsyncCopyCallback function.
